crates/chia-consensus/src/gen/condition_tools.rs

```diff
@@ -0,0 +1,86 @@
+use crate::consensus_constants::ConsensusConstants;
+use crate::gen::opcodes::{
+    ConditionOpcode, AGG_SIG_AMOUNT, AGG_SIG_ME, AGG_SIG_PARENT, AGG_SIG_PARENT_AMOUNT,
+    AGG_SIG_PARENT_PUZZLE, AGG_SIG_PUZZLE, AGG_SIG_PUZZLE_AMOUNT,
+};
+use crate::gen::owned_conditions::OwnedSpend;
+use chia_protocol::Bytes;
+use chia_protocol::Coin;
+
+pub fn make_aggsig_final_message(
+    opcode: ConditionOpcode,
+    msg: &[u8],
+    spend: &OwnedSpend,
+    constants: &ConsensusConstants,
+) -> Vec<u8> {
+    let mut result = msg.to_vec();
+    result.extend(match opcode {
+        AGG_SIG_PARENT => [
+            spend.parent_id.as_slice(),
+            constants.agg_sig_parent_additional_data.as_slice(),
+        ]
+        .concat(),
+        AGG_SIG_PUZZLE => [
+            spend.puzzle_hash.as_slice(),
+            constants.agg_sig_puzzle_additional_data.as_slice(),
+        ]
+        .concat(),
+        AGG_SIG_AMOUNT => [
+            u64_to_bytes(spend.coin_amount).as_slice(),
+            constants.agg_sig_amount_additional_data.as_slice(),
+        ]
+        .concat(),
+        AGG_SIG_PUZZLE_AMOUNT => [
+            spend.puzzle_hash.as_slice(),
+            u64_to_bytes(spend.coin_amount).as_slice(),
+            constants.agg_sig_puzzle_amount_additional_data.as_slice(),
+        ]
+        .concat(),
+        AGG_SIG_PARENT_AMOUNT => [
+            spend.parent_id.as_slice(),
+            u64_to_bytes(spend.coin_amount).as_slice(),
+            constants.agg_sig_parent_amount_additional_data.as_slice(),
+        ]
+        .concat(),
+        AGG_SIG_PARENT_PUZZLE => [
+            spend.parent_id.as_slice(),
+            spend.puzzle_hash.as_slice(),
+            constants.agg_sig_parent_puzzle_additional_data.as_slice(),
+        ]
+        .concat(),
+        AGG_SIG_ME => {
+            let coin: Coin = Coin::new(spend.parent_id, spend.puzzle_hash, spend.coin_amount);
+            [
+                coin.coin_id().as_slice(),
+                constants.agg_sig_me_additional_data.as_slice(),
+            ]
+            .concat()
+        }
+        _ => Vec::<u8>::new(),
+    });
+
+    result
+}
+
+pub fn u64_to_bytes(val: u64) -> Bytes {
```

Comment:
this should have a unit test

Reply:
I am very worried about conversions like this here and in my own code. I think we should have extensive fuzz testing to compare the results against the Allocator too. Maybe out of scope of this PR though

Reply:
The main concern is leading zeros being added or not, for example

Reply:
I agree and will implement some unit tests, but this code was adapted from `Coin.coin_id()` which is used in many critical places

Reply:
Yeah if it's the same, probably not too urgent but better test coverage of that existing code would be good too in the future